File: backend/app/routes/registration.py
"""
Registration routes for driver registration API endpoints.
Handles driver registration requests.
"""

from typing import Optional
from fastapi import APIRouter, Depends, File, Form, UploadFile, HTTPException, status
from datetime import date
from app.middleware.api_key_auth import verify_api_key
from app.schemas.driver_schema import (
    DriverRegistrationRequest,
    DriverRegistrationResponse,
    RegistrationData,
    ErrorResponse
)
from app.services.driver_service import DriverService


router = APIRouter()


@router.post(
    "/driver",
    response_model=DriverRegistrationResponse,
    status_code=status.HTTP_201_CREATED,
    responses={
        400: {"model": ErrorResponse, "description": "Validation error"},
        409: {"model": ErrorResponse, "description": "Email already exists"},
        401: {"model": ErrorResponse, "description": "Unauthorized"},
        500: {"model": ErrorResponse, "description": "Internal server error"}
    },
    dependencies=[Depends(verify_api_key)]
)
async def register_driver(
    # Required fields
    email: str = Form(...),
    auth_provider: str = Form(...),
    first_name: str = Form(...),
    surname: str = Form(...),
    birthdate: str = Form(...),
    birthplace: str = Form(...),
    license_number: str = Form(...),
    license_expiry_date: str = Form(...),
    address_line_1: str = Form(...),
    primary_phone: str = Form(...),
    # Optional fields
    oauth_id: Optional[str] = Form(None),
    middle_name: Optional[str] = Form(None),
    address_line_2: Optional[str] = Form(None),
    secondary_phone: Optional[str] = Form(None),
    profile_photo: Optional[UploadFile] = File(None),
):
    """
    Register a new driver.

    Creates a new driver registration with status 'pending'.
    Requires API key authentication via X-API-Key header.

    Args:
        All driver registration fields as form data
        profile_photo: Optional profile photo file (JPG/PNG, max 5MB)

    Returns:
        DriverRegistrationResponse: Success response with driver info

    Raises:
        HTTPException: Various error codes based on validation or conflicts
    """
    try:
        # Parse dates
        try:
            birthdate_obj = date.fromisoformat(birthdate)
            license_expiry_obj = date.fromisoformat(license_expiry_date)
        except ValueError as e:
            raise HTTPException(
                status_code=400,
                detail=f"Invalid date format. Use YYYY-MM-DD: {str(e)}"
            )

        # Create validated request object
        driver_request = DriverRegistrationRequest(
            email=email,
            auth_provider=auth_provider,
            first_name=first_name,
            middle_name=middle_name,
            surname=surname,
            birthdate=birthdate_obj,
            birthplace=birthplace,
            license_number=license_number,
            license_expiry_date=license_expiry_obj,
            address_line_1=address_line_1,
            address_line_2=address_line_2,
            primary_phone=primary_phone,
            secondary_phone=secondary_phone,
            oauth_id=oauth_id
        )

        # Create driver registration
        result = await DriverService.create_driver_registration(
            driver_data=driver_request,
            profile_photo=profile_photo
        )

        # Return success response
        return DriverRegistrationResponse(
            success=True,
            message="Registration submitted successfully",
            data=RegistrationData(**result)
        )

    except HTTPException:
        # Re-raise HTTP exceptions
        raise
    except ValueError as e:
        # Validation errors from Pydantic
        raise HTTPException(
            status_code=400,
            detail=str(e)
        )
    except Exception as e:
        # Unexpected errors
        print(f"Error in register_driver: {str(e)}")
        raise HTTPException(
            status_code=500,
            detail="An error occurred while processing your registration. Please try again."
        )


@router.get("/health")
async def health_check():
    """Health check endpoint for registration service."""
    return {"status": "healthy", "service": "Registration API"}

File: types/registration.ts
/**
 * TypeScript types for driver registration
 */

export type AuthProvider = 'email' | 'google' | 'apple';

export interface RegistrationFormData {
  // Authentication
  email: string;
  authProvider: AuthProvider;
  oauthId?: string;

  // Personal Information
  firstName: string;
  middleName?: string;
  surname: string;
  birthdate: string; // YYYY-MM-DD format
  birthplace: string;

  // License Information
  licenseNumber: string;
  licenseExpiryDate: string; // YYYY-MM-DD format

  // Contact Information
  addressLine1: string;
  addressLine2?: string;
  primaryPhone: string;
  secondaryPhone?: string;

  // Profile Photo
  profilePhoto?: {
    uri: string;
    type: string;
    name: string;
  };
}

export interface ProfilePhoto {
  uri: string;
  type: string;
  name: string;
  size?: number;
}

export interface RegistrationApiResponse {
  success: boolean;
  message: string;
  data?: {
    driver_id: string;
    email: string;
    status: string;
    created_at: string;
  };
  error?: string;
  details?: Array<{
    field: string;
    message: string;
  }>;
}

export interface OAuthUserInfo {
  email: string;
  firstName?: string;
  surname?: string;
  profilePhoto?: string;
}

export interface GoogleOAuthResponse {
  type: 'success' | 'error' | 'cancel';
  authentication?: {
    accessToken: string;
  };
  params?: {
    access_token: string;
  };
}

export interface AppleOAuthResponse {
  email?: string;
  fullName?: {
    givenName?: string;
    familyName?: string;
  };
}

// Authentication types
export interface LoginCredentials {
  emailOrUsername: string;
  password: string;
}

export interface LoginResponse {
  success: boolean;
  message: string;
  data?: {
    driver_id: string;
    email: string;
    username?: string;
    first_name: string;
    surname: string;
    status: string;
    access_token: string;
    token_type: string;
  };
  error?: string;
  detail?: string;
}

export interface DriverData {
  driver_id: string;
  email: string;
  username?: string;
  firstName: string;
  surname: string;
  status: string;
  accessToken: string;
}
